File: src/cpp/win32/HwndUtil.cpp
#include "HwndUtil.h"

#include <psapi.h>
#include <wil/resource.h>

#include <cwctype>
#include <chrono>
#include <mutex>
#include <string_view>
#include <unordered_map>

#pragma comment(lib, "psapi.lib")

namespace imeindicator::win32 {
namespace {

struct CacheEntry {
    std::wstring name;
    std::chrono::steady_clock::time_point expiresAt;
};

std::mutex g_cacheMutex;
std::unordered_map<DWORD, CacheEntry> g_processNameCache;
constexpr auto kCacheTtl = std::chrono::seconds(1);

std::wstring extractFileNameFromPath(std::wstring_view path)
{
    auto slash = path.find_last_of(L"\\/");
    if (slash == std::wstring_view::npos) {
        return std::wstring(path);
    }
    return std::wstring(path.substr(slash + 1));
}

} // namespace

std::wstring getProcessNameByPid(DWORD pid)
{
    if (pid == 0) return {};

    {
        std::lock_guard lk(g_cacheMutex);
        auto it = g_processNameCache.find(pid);
        if (it != g_processNameCache.end()) {
            if (std::chrono::steady_clock::now() < it->second.expiresAt) {
                return it->second.name;
            }
            g_processNameCache.erase(it);
        }
    }

    wil::unique_handle hProcess(::OpenProcess(
        PROCESS_QUERY_LIMITED_INFORMATION, FALSE, pid));
    if (!hProcess) {
        return {};
    }

    wchar_t buffer[MAX_PATH] = {};
    DWORD size = MAX_PATH;
    if (!::QueryFullProcessImageNameW(hProcess.get(), 0, buffer, &size) || size == 0) {
        return {};
    }

    std::wstring name = extractFileNameFromPath(std::wstring_view(buffer, size));

    {
        std::lock_guard lk(g_cacheMutex);
        g_processNameCache[pid] = CacheEntry{
            name,
            std::chrono::steady_clock::now() + kCacheTtl
        };
    }
    return name;
}

std::wstring getProcessNameByHwnd(HWND hwnd)
{
    if (!hwnd) return {};
    DWORD pid = 0;
    ::GetWindowThreadProcessId(hwnd, &pid);
    return getProcessNameByPid(pid);
}

std::wstring getWindowTitle(HWND hwnd)
{
    if (!hwnd) return {};
    const int len = ::GetWindowTextLengthW(hwnd);
    if (len <= 0) return {};
    std::wstring buf(static_cast<size_t>(len) + 1, L'\0');
    const int got = ::GetWindowTextW(hwnd, buf.data(), len + 1);
    if (got <= 0) return {};
    buf.resize(static_cast<size_t>(got));
    return buf;
}

std::wstring getWindowClassName(HWND hwnd)
{
    if (!hwnd) return {};
    wchar_t buf[256] = {};
    const int got = ::GetClassNameW(hwnd, buf, static_cast<int>(std::size(buf)));
    if (got <= 0) return {};
    return std::wstring(buf, static_cast<size_t>(got));
}

HKL getKeyboardLayoutOfHwnd(HWND hwnd)
{
    if (!hwnd) return ::GetKeyboardLayout(0);
    DWORD pid = 0;
    DWORD tid = ::GetWindowThreadProcessId(hwnd, &pid);
    return ::GetKeyboardLayout(tid);
}

bool isJapaneseHkl(HKL hkl)
{
    // HKL の下位 16 bit が言語 ID。0x0411 = 日本語(ja-JP)。
    constexpr WORD kJapaneseLangId = 0x0411;
    return LOWORD(reinterpret_cast<uintptr_t>(hkl)) == kJapaneseLangId;
}

// ===== HotkeyP マージ (010-hotkeyp-merge) で追加されたヘルパー =====

std::wstring getForegroundExeName()
{
    HWND hw = GetForegroundWindow();
    if (!hw) return {};
    return getProcessNameByHwnd(hw);
}

namespace {

struct FindWindowContext {
    std::wstring targetExe;  // 比較対象（ファイル名のみ、小文字化済み）
    HWND found = nullptr;
};

std::wstring toLowerCopy(std::wstring_view s) {
    std::wstring r(s);
    for (auto& c : r) c = static_cast<wchar_t>(::towlower(c));
    return r;
}

BOOL CALLBACK enumWindowsForExe(HWND hwnd, LPARAM lParam)
{
    auto* ctx = reinterpret_cast<FindWindowContext*>(lParam);

    // 表示可能なトップレベルウィンドウのみ対象（HotkeyP 元 findWindow 相当）
    if (!IsWindowVisible(hwnd)) return TRUE;
    if (GetWindow(hwnd, GW_OWNER) != nullptr) return TRUE;  // 所有者ありはサブウィンドウ
    if (GetWindowTextLengthW(hwnd) == 0) return TRUE;        // タイトル空はダイアログ等

    auto name = getProcessNameByHwnd(hwnd);
    if (name.empty()) return TRUE;

    if (toLowerCopy(name) == ctx->targetExe) {
        ctx->found = hwnd;
        return FALSE;  // 列挙終了
    }
    return TRUE;
}

} // namespace

HWND findWindowByExeName(std::wstring_view exeFullPathOrName)
{
    if (exeFullPathOrName.empty()) return nullptr;

    // 末尾のファイル名部分のみ抽出
    auto fileName = extractFileNameFromPath(exeFullPathOrName);

    FindWindowContext ctx;
    ctx.targetExe = toLowerCopy(fileName);

    EnumWindows(enumWindowsForExe, reinterpret_cast<LPARAM>(&ctx));
    return ctx.found;
}

bool bringWindowToFront(HWND hwnd) noexcept
{
    if (!hwnd || !IsWindow(hwnd)) return false;

    // 最小化状態なら復元
    if (IsIconic(hwnd)) {
        ShowWindow(hwnd, SW_RESTORE);
    }

    // SetForegroundWindow は通常、フォアグラウンドロックの制約があり呼び出し失敗するケースあり
    // AttachThreadInput で一時的にスレッド入力をアタッチして強制前面化（HotkeyP 互換）
    DWORD foreThread = GetWindowThreadProcessId(GetForegroundWindow(), nullptr);
    DWORD targetThread = GetWindowThreadProcessId(hwnd, nullptr);
    DWORD currentThread = GetCurrentThreadId();

    if (foreThread != currentThread) {
        AttachThreadInput(currentThread, foreThread, TRUE);
    }
    if (targetThread != currentThread) {
        AttachThreadInput(currentThread, targetThread, TRUE);
    }

    BringWindowToTop(hwnd);
    BOOL result = SetForegroundWindow(hwnd);

    if (foreThread != currentThread) {
        AttachThreadInput(currentThread, foreThread, FALSE);
    }
    if (targetThread != currentThread) {
        AttachThreadInput(currentThread, targetThread, FALSE);
    }

    return result != FALSE;
}

} // namespace imeindicator::win32

File: src/csharp/IMEIndicator/Services/HwndUtil.cs
// Copyright (C) 2026 IMEIndicator Project
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License v2 or later.
// See COPYING in the repository root for the full license text.

using System.Text;
using IMEIndicator.Interop;

namespace IMEIndicator.Services;

/// <summary>
/// ウィンドウ・プロセス名解決のヘルパー。移植元: src/cpp/win32/HwndUtil.h / .cpp のうち、
/// <c>HotkeyService</c>（T068）の multInst=false 判定が必要とする
/// <c>findWindowByExeName</c> / <c>bringWindowToFront</c> のみを移植する（<c>getWindowTitle</c> /
/// <c>getKeyboardLayoutOfHwnd</c> / <c>isJapaneseHkl</c> 等、C++ 版の他の用途向け関数は対象外）。
/// </summary>
internal static class HwndUtil
{
    private static readonly Lock CacheLock = new();
    private static readonly Dictionary<uint, (string Name, DateTime ExpiresAt)> ProcessNameCache = [];
    private static readonly TimeSpan CacheTtl = TimeSpan.FromSeconds(1);

    /// <summary>
    /// 指定した実行ファイル名（フルパスでも可、末尾のファイル名部分のみ比較）を持つプロセスの
    /// 可視トップレベルウィンドウを探す。移植元 <c>findWindowByExeName</c>。
    /// </summary>
    /// <param name="exeFullPathOrName">比較対象の実行ファイルパスまたはファイル名。</param>
    /// <returns>見つかった場合はそのウィンドウハンドル、無ければ 0。</returns>
    public static nint FindWindowByExeName(string exeFullPathOrName)
    {
        if (string.IsNullOrEmpty(exeFullPathOrName))
        {
            return 0;
        }

        string target = ExtractFileName(exeFullPathOrName).ToLowerInvariant();
        nint found = 0;

        bool EnumProc(nint hwnd, nint lParam)
        {
            // 表示可能なトップレベルウィンドウのみ対象（HotkeyP 元 findWindow 相当）。
            if (!NativeMethods.IsWindowVisible(hwnd))
            {
                return true;
            }

            if (NativeMethods.GetWindow(hwnd, NativeConstants.GW_OWNER) != 0)
            {
                return true; // 所有者ありはサブウィンドウ
            }

            if (NativeMethods.GetWindowTextLengthW(hwnd) == 0)
            {
                return true; // タイトル空はダイアログ等
            }

            string name = GetProcessNameByHwnd(hwnd);
            if (name.Length == 0)
            {
                return true;
            }

            if (string.Equals(name, target, StringComparison.OrdinalIgnoreCase))
            {
                found = hwnd;
                return false; // 列挙終了
            }

            return true;
        }

        NativeMethods.EnumWindows(EnumProc, 0);
        return found;
    }

    /// <summary>
    /// 指定ウィンドウを最小化解除・前面化する。フォアグラウンドロックの制約を
    /// <c>AttachThreadInput</c> による一時アタッチで回避する（HotkeyP 互換）。移植元 <c>bringWindowToFront</c>。
    /// </summary>
    public static bool BringWindowToFront(nint hwnd)
    {
        if (hwnd == 0 || !NativeMethods.IsWindow(hwnd))
        {
            return false;
        }

        if (NativeMethods.IsIconic(hwnd))
        {
            NativeMethods.ShowWindow(hwnd, NativeConstants.SW_RESTORE);
        }

        uint foreThread = NativeMethods.GetWindowThreadProcessId(NativeMethods.GetForegroundWindow(), out _);
        uint targetThread = NativeMethods.GetWindowThreadProcessId(hwnd, out _);
        uint currentThread = NativeMethods.GetCurrentThreadId();

        if (foreThread != currentThread)
        {
            NativeMethods.AttachThreadInput(currentThread, foreThread, true);
        }

        if (targetThread != currentThread)
        {
            NativeMethods.AttachThreadInput(currentThread, targetThread, true);
        }

        NativeMethods.BringWindowToTop(hwnd);
        bool result = NativeMethods.SetForegroundWindow(hwnd);

        if (foreThread != currentThread)
        {
            NativeMethods.AttachThreadInput(currentThread, foreThread, false);
        }

        if (targetThread != currentThread)
        {
            NativeMethods.AttachThreadInput(currentThread, targetThread, false);
        }

        return result;
    }

    private static string GetProcessNameByHwnd(nint hwnd)
    {
        if (hwnd == 0)
        {
            return string.Empty;
        }

        NativeMethods.GetWindowThreadProcessId(hwnd, out uint pid);
        return GetProcessNameByPid(pid);
    }

    // PID → 実行ファイル名（末尾部分のみ）。1 秒 TTL でキャッシュする（移植元と同じ）。
    // 1 回の FindWindowByExeName（EnumWindows）呼び出し中に同一プロセスの複数ウィンドウを
    // 引くケース（ブラウザ・IDE 等）で OpenProcess/QueryFullProcessImageNameW の呼び出しを
    // 減らすための最適化。ホットキー押下という同期経路で呼ばれるため、体感遅延を避ける目的。
    private static string GetProcessNameByPid(uint pid)
    {
        if (pid == 0)
        {
            return string.Empty;
        }

        lock (CacheLock)
        {
            if (ProcessNameCache.TryGetValue(pid, out (string Name, DateTime ExpiresAt) cached))
            {
                if (DateTime.UtcNow < cached.ExpiresAt)
                {
                    return cached.Name;
                }

                ProcessNameCache.Remove(pid);
            }
        }

        nint hProcess = NativeMethods.OpenProcess(NativeConstants.PROCESS_QUERY_LIMITED_INFORMATION, false, pid);
        if (hProcess == 0)
        {
            return string.Empty;
        }

        try
        {
            var buffer = new StringBuilder(260); // MAX_PATH
            uint size = (uint)buffer.Capacity;
            if (!NativeMethods.QueryFullProcessImageNameW(hProcess, 0, buffer, ref size) || size == 0)
            {
                return string.Empty;
            }

            string name = ExtractFileName(buffer.ToString(0, (int)size));

            lock (CacheLock)
            {
                ProcessNameCache[pid] = (name, DateTime.UtcNow + CacheTtl);
            }

            return name;
        }
        finally
        {
            NativeMethods.CloseHandle(hProcess);
        }
    }

    private static string ExtractFileName(string path)
    {
        int slash = path.LastIndexOfAny(['\\', '/']);
        return slash < 0 ? path : path[(slash + 1)..];
    }
}
